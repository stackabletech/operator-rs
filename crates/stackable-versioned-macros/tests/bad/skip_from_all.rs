use stackable_versioned_macros::versioned;

fn main() {
    #[versioned(
        version(name = "v1alpha1"),
        version(name = "v1beta1"),
        version(name = "v1"),
        options(skip(from))
    )]
    pub struct Foo {
        #[versioned(
            added(since = "v1beta1"),
            deprecated(since = "v1", note = "not needed")
        )]
        deprecated_bar: usize,
        baz: bool,
    }

    let foo_v1alpha1 = v1alpha1::Foo { baz: true };

    // There are no From impls for any version. You need to convert it manually.
    #[allow(dead_code)]
    let foo_v1beta1 = v1beta1::Foo::from(foo_v1alpha1);
}
