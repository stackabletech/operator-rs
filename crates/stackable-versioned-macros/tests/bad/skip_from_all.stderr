error[E0308]: mismatched types
  --> tests/bad/skip_from_all.rs:23:42
   |
23 |     let foo_v1beta1 = v1beta1::Foo::from(foo_v1alpha1);
   |                       ------------------ ^^^^^^^^^^^^ expected `v1beta1::Foo`, found `v1alpha1::Foo`
   |                       |
   |                       arguments to this function are incorrect
   |
   = note: `v1alpha1::Foo` and `v1beta1::Foo` have similar names, but are actually distinct types
note: `v1alpha1::Foo` is defined in module `crate::main::v1alpha1` of the current crate
  --> tests/bad/skip_from_all.rs:4:5
   |
4  | /     #[versioned(
5  | |         version(name = "v1alpha1"),
6  | |         version(name = "v1beta1"),
7  | |         version(name = "v1"),
8  | |         options(skip(from))
9  | |     )]
   | |______^
note: `v1beta1::Foo` is defined in module `crate::main::v1beta1` of the current crate
  --> tests/bad/skip_from_all.rs:4:5
   |
4  | /     #[versioned(
5  | |         version(name = "v1alpha1"),
6  | |         version(name = "v1beta1"),
7  | |         version(name = "v1"),
8  | |         options(skip(from))
9  | |     )]
   | |______^
note: associated function defined here
  --> $RUST/core/src/convert/mod.rs
   |
   |     fn from(value: T) -> Self;
   |        ^^^^
   = note: this error originates in the attribute macro `versioned` (in Nightly builds, run with -Z macro-backtrace for more info)
